packages/react-router/src/path.ts

```diff
@@ -8,7 +8,13 @@ export interface Segment {
 }
 
 export function joinPaths(paths: Array<string | undefined>) {
-  return cleanPath(paths.filter(Boolean).join('/'))
+  return cleanPath(
+    paths
+      .filter((val) => {
```

Comment:
This was `Boolean` filtering was causing falsey number values (like `0`, `-1`, etc...) to be excluded out of the joined path.

This ripples out into the router logic, where all the right routes don't get matched since  the pathname it's using for checking will be incomplete or in some cases even outright incorrect.